schema/ksl-schema.kdl

```diff
@@ -0,0 +1,925 @@
+                Only used for numeric values. Constrains them to be greater than
+                the given number.
+                """ {
+                arg {
+                    type number
+                }
+            }
+            node gte about="""
+                Greater than or equal to
+
+                Only used for numeric values. Constrains them to be greater than
+                or equal to the given number.
+                """ {
+                arg {
+                    type number
+                }
+            }
+            node lt about="""
+                Less than
+
+                Only used for numeric values. Constrains them to be less than
+                the given number.
+                """ {
+                arg {
+                    type number
+                }
+            }
+            node lte about="""
+                Less than or equal to
+
+                Only used for numeric values. Constrains them to be less than or
+                equal to the given number
+                """ {
+                arg {
+                    type number
+                }
+            }
+        }
+    }
+    // Validations shared across all types.
+    node shared-validations id=shared-validations {
+        children {
+            node type about="The type for this value\n\nMultiple arguments signify a sum type." {
+                repeatable
+                args {
+                    min 1
+                    type string
+                    enum string boolean number integer #null
+                    distinct
+                }
+            }
+            // TODO: establish equality expectations.
+            node const about="""
+                Specific value
+
+                Exact value that this value must match. Equivalent to a
+                single-value `enum` validation.
+                """ {
+                arg about="Constant value"
+            }
+            node enum about="""
+                Enumeration of values
+
+                An enumeration of possible values
+                """ {
+                repeatable
+                args about="Enumeration choices" {
+                    min 1
+                }
+                children {
+                    node allow-others about="""
+                        Allow other choices
+
+                        Whether other values than those explicitly enumerated
+                        may be provided, so long as they pass other validations
+                        in the node.
+
+                        While apparently redundant, this option may be useful in
+                        cases where there's a set of suggested values, but
+                        others are acceptable. This information can then be used
+                        by tooling to e.g. suggest completion items.
+                        """ {
+                        arg {
+                            type boolean
+                            default #true
```

Comment:
Even if we don't remove defaults, as I think we should, I think this should be `#false`?

Reply:
It might seem confusing, but the logic is something like:

1. no `allow-others` node: the value is false, because there’s no value at all
2. An `allow-others` node with no argument: the value is `#true`, because we have the node, then look up the value, the value is missing, therefore we use the default
3. `allow-others #false`: forces the value to false.

the reason for doing things this way is that so nodes that work this way “present or not” can be overridden when using `ref`. So if you’re “inheriting” a node that includes `allow-others`, you can turn off that default